pkg/s3-proxy/config/validation.go

```diff
@@ -268,26 +269,75 @@ func validateSSLConfig(serverSSL *ServerSSLConfig, section string) error {
 
 	for i := range serverSSL.Certificates {
 		cert := &serverSSL.Certificates[i]
-		if cert.Certificate == nil {
-			if cert.CertificateURL == nil {
-				return fmt.Errorf(
-					"%s.ssl.certificates[%d] must have either certificate or certificateUrl set", section,
-					i)
+		err := validateSSLCertificateComponentConfig(
+			cert.Certificate, cert.CertificateURL, cert.CertificateURLConfig,
+			fmt.Sprintf("%s.ssl.certificates[%d].certificate", section, i))
+
+		if err != nil {
+			return err
+		}
+
+		err = validateSSLCertificateComponentConfig(
+			cert.PrivateKey, cert.PrivateKeyURL, cert.PrivateKeyURLConfig,
+			fmt.Sprintf("%s.ssl.certificates[%d].privateKey", section, i))
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func validateSSLCertificateComponentConfig(
+	component *string, componentURL *string, componentURLConfig *SSLURLConfig, componentName string) error {
+	if component == nil {
+		if componentURL == nil {
+			return fmt.Errorf("either %s or %sUrl must be set", componentName, componentName)
+		}
+
+		err := utils.ValidateDocumentURL(*componentURL)
+		if err != nil {
+			return fmt.Errorf("%sUrl is a malformed/unsupported URL: %w: %s", componentName, err, *componentURL)
```

Comment:
> Moreover, there is a test that seems to be failing every time:
> 
> ```
> --- FAIL: TestTLSServer/Certificate/PrivateKey_stored_in_S3 
> ```
> 
> There is the fail: `tls_integration_test.go:606: Expected no error but got Get "[127.0.0.1:8084/mount/folder1/test.txt](https://127.0.0.1:8084/mount/folder1/test.txt)": EOF` Source: https://github.com/oxyno-zeta/s3-proxy/runs/5496535471?check_suite_focus=true#step:9:1315
> 
> PS: I already run it twice to be sure that it isn't just flaky. The other fail is a flaky test, please ignore it

Not seeing that on my end... hrm. What platform are you on? EOF being returned from httpClient is not normal behavior.

Reply:
> Not seeing that on my end... hrm. What platform are you on? EOF being returned from httpClient is not normal behavior.

It is on the Github CI node which on "ubuntu@latest".

Reply:
Still looking into this.

Reply:
Interesting... this line (tls_test.go:336) refuses to compile on Ubuntu 20.04:
```
t.Errorf("expected nil result but got tls.Config %#v", *tlsConfig)
```

Reply:
Ok, that was a red herring. The real issue was port assignment. That test just happened to be (repeatably) assigned port 8088, which conflicted with another port assignment being made in the CI workflow.

Latest commit fixes this (and updates the Go version to 1.17.8, which I suspected might be the issue but was another non-issue; but I figure it can't hurt to be on the latest patch level).